compiler/src/iree/compiler/ExternalInterfaces/UtilExternalModels.cpp

```diff
@@ -46,6 +48,229 @@ getDivisibilityOfOperand(Value v,
+      return divisibilityMap[expr];
+    }
+    return IREE::Util::IntegerDivisibility::getMinDivisibility().getValue();
+  }
+
+  /// The divisibility of an addition is the GCD of its constituents'
+  /// divisibilities. This callback is used for subtraction as well.
+  IREE::Util::ConstantIntDivisibility visitAddExpr(AffineBinaryOpExpr expr) {
+    if (divisibilityMap.contains(expr)) {
+      return divisibilityMap[expr];
+    }
+    IREE::Util::ConstantIntDivisibility lhsDiv = visit(expr.getLHS());
+    IREE::Util::ConstantIntDivisibility rhsDiv = visit(expr.getRHS());
+    return lhsDiv.getUnion(rhsDiv);
+  }
+
+  /// The divisibility of a multiplication is the product of its constituents'
+  /// divisibilities.
+  IREE::Util::ConstantIntDivisibility visitMulExpr(AffineBinaryOpExpr expr) {
+    if (divisibilityMap.contains(expr)) {
+      return divisibilityMap[expr];
+    }
+    IREE::Util::ConstantIntDivisibility lhsDiv = visit(expr.getLHS());
+    IREE::Util::ConstantIntDivisibility rhsDiv = visit(expr.getRHS());
+    return IREE::Util::ConstantIntDivisibility(lhsDiv.udiv() * rhsDiv.udiv(),
+                                               lhsDiv.sdiv() * rhsDiv.sdiv());
+  }
+
+  IREE::Util::ConstantIntDivisibility
+  visitFloorDivExpr(AffineBinaryOpExpr expr) {
+    return visitDivExpr(expr);
+  }
+
+  IREE::Util::ConstantIntDivisibility
+  visitCeilDivExpr(AffineBinaryOpExpr expr) {
+    return visitDivExpr(expr);
+  }
+
+  /// Mod expressions could be inferred to be zero in some cases, but for now
+  /// just return the minimum divisibility.
+  /// TODO(Max191): Handle evenly divisible cases, and ensure that the zero
+  /// divisibility propagates properly through parent expressions.
+  IREE::Util::ConstantIntDivisibility visitModExpr(AffineBinaryOpExpr expr) {
+    return visitInvalidExpr(expr);
+  }
+
+private:
+  IREE::Util::ConstantIntDivisibility
+  visitInvalidExpr(AffineBinaryOpExpr expr) {
+    return IREE::Util::IntegerDivisibility::getMinDivisibility().getValue();
+  }
+
+  /// Helper shared by ceildiv and floordiv implementations. The divisibility of
+  /// a division is simply the quotient of its constituents' divisibilities as
+  /// long as the division has no remainder. If there is a remainder, then the
+  /// divisibility cannot be easily inferred, so we fallback to the minimum
+  /// divisibility.
+  IREE::Util::ConstantIntDivisibility visitDivExpr(AffineBinaryOpExpr expr) {
+    if (divisibilityMap.contains(expr)) {
+      return divisibilityMap[expr];
+    }
+    IREE::Util::ConstantIntDivisibility lhsDiv = visit(expr.getLHS());
+    auto constRhs = dyn_cast<AffineConstantExpr>(expr.getRHS());
+    if (!constRhs) {
+      return IREE::Util::ConstantIntDivisibility(1, 1);
+    }
+    int64_t constValue = constRhs.getValue();
+    uint64_t divUDiv = lhsDiv.udiv() % static_cast<uint64_t>(constValue) == 0
+                           ? lhsDiv.udiv() / static_cast<uint64_t>(constValue)
+                           : 1;
```

Comment:
`static_cast<uint64_t>` on negative number creates a very large positive number. E.g., `-4` becomes `18446744073709551612`.

I'd initialize the value to `1` and create few if statement to handle the case.